app/lib/payment_gateways/stripe_crypt.rb

```diff
@@ -24,12 +24,14 @@ def update!(payment_method_id)
     end
 
     def create_stripe_setup_intent
-      setup_intent_params = {
-        payment_method_types: ['card'],
-        usage: 'off_session',
-        customer: customer.id
-      }
-      Stripe::SetupIntent.create(setup_intent_params, api_key)
+      if customer.present?
+        setup_intent_params = {
+          payment_method_types: ['card'],
+          usage: 'off_session',
+          customer: customer.id
+        }
+        Stripe::SetupIntent.create(setup_intent_params, api_key)
+      end
```

Comment:
Why do you need the new `if`? In my local tests, when `find_or_create_customer` catches the exception and returns normally, then `customer.present?` returns true. So this `if` doesn't add any protection.

Reply:
Hm, that's weird. I think when I tested it, when `Stripe::Customer.retrieve` was throwing `Stripe::InvalidRequestError`, the `find_or_create_customer` returned `nil`.

Reply:
I agree it's a weird behavior, that's why I checked out the branch and tried again, assuming there was something wrong with my server the last time. But it failed again. Check this debugger log (breakpoint at [stripe_crypt.rb:27](https://github.com/3scale/porta/pull/3351/files#diff-4c16dc8fb0816a60d4675d1987c5588b66341daf5edd80d0cb2fb1ececb5a9b6R27)):

```
0> customer
  PaymentDetail Load (0.7ms)  SELECT  `payment_details`.* FROM `payment_details` WHERE `payment_details`.`account_id` = 3 ORDER BY `payment_details`.`id` DESC LIMIT 1
  ↳ app/models/account/payment_details.rb:29
An invalid request occurred.
=> true

0> customer.present
=> undefined method `present' for true:TrueClass

0> customer.present?
=> true

0> Rails.logger
=> #<ActiveSupport::Logger:0x000000000ceeab18 @level=0, @progname=nil, @default_formatter=#<Logger::Formatter:0x000000000ceeaaa0 @datetime_format=nil>, @formatter=#<System::ErrorReporting::LogFormatter:0x000000000ac7d4e0 @datetime_format=nil, @thread_key="activesupport_tagged_logging_tags:90434160">, @logdev=#<Logger::LogDevice:0x000000000ceeaa50 @shift_period_suffix=nil, @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<File:/home/jlledom/redhat/porta/log/development.log>, @mon_mutex=#<Thread::Mutex:0x000000000ceeaa00>, @mon_mutex_owner_object_id=108483880, @mon_owner=nil, @mon_count=0>, @local_levels=#<Concurrent::Map:0x000000000ceea988 entries=0 default_proc=nil>>

0> Rails.logger.warn("test")
test
=> true
```

`Rails.logger.warn` returns `true`. Maybe we are not using the same logger class? Check my output for `Rails.logger`.

Reply:
Ah,OK, I just realized that I tested **before** this commit https://github.com/3scale/porta/pull/3351/commits/48d5e7c199e4b4a4166b381590443a5092de1c61 :upside_down_face: 

the method used to return `nil` but after switching from `puts` to the logger, it now returns `true` indeed :upside_down_face: 

Anyway, I also think we need to unify the logic, and handle the "not found" and "deleted" Stripe customer in the same way (i.e. creating a new customer)

Reply:
@jlledom and @mayorova 

so should we create new customer if customer is not present?

Reply:
> so should we create new customer if customer is not present?

I think so, after that, you won't need to add that `if customer.present?` check. Also, you won't need the `&` at https://github.com/3scale/porta/pull/3351/files#diff-46e8193f2a380a515bbe6ef58c221d0498d97749113ebf56fa41823352e02fbfR48